packages/sanity/src/core/form/store/formState.ts

```diff
@@ -128,937 +191,1106 @@ function isChangedValue(value: any, comparisonValue: any) {
   return !_isEqual(value, comparisonValue)
 }
 
-/*
- * Takes a field in context of a parent object and returns prepared props for it
- */
-function prepareFieldMember(props: {
-  field: ObjectField
-  parent: RawState<ObjectSchemaType, unknown> & {
-    groups: FormFieldGroup[]
-    selectedGroup: FormFieldGroup
-  }
-  index: number
-}): ObjectMember | HiddenField | null {
-  const {parent, field, index} = props
-  const fieldPath = pathFor([...parent.path, field.name])
-  const fieldLevel = getFieldLevel(field.type, parent.level + 1)
-
-  const parentValue = parent.value
-  const parentComparisonValue = parent.comparisonValue
-  if (!isAcceptedObjectValue(parentValue)) {
-    // Note: we validate each field, before passing it recursively to this function so getting this error means that the
-    // ´prepareFormState´ function itself has been called with a non-object value
-    throw new Error('Unexpected non-object value')
+export interface CreatePrepareFormStateOptions {
+  decorators?: {
+    prepareFieldMember?: FunctionDecorator<PrepareFieldMember>
+    prepareObjectInputState?: FunctionDecorator<PrepareObjectInputState>
+    prepareArrayOfPrimitivesInputState?: FunctionDecorator<PrepareArrayOfPrimitivesInputState>
+    prepareArrayOfObjectsInputState?: FunctionDecorator<PrepareArrayOfObjectsInputState>
+    prepareArrayOfObjectsMember?: FunctionDecorator<PrepareArrayOfObjectsMember>
+    prepareArrayOfPrimitivesMember?: FunctionDecorator<PrepareArrayOfPrimitivesMember>
+    preparePrimitiveInputState?: FunctionDecorator<PreparePrimitiveInputState>
   }
+}
 
-  const normalizedFieldGroupNames = field.group ? castArray(field.group) : []
-  const inSelectedGroup = isFieldEnabledByGroupFilter(
-    parent.groups,
-    field.group,
-    parent.selectedGroup,
-  )
+export interface RootFormStateOptions {
+  schemaType: ObjectSchemaType
+  documentValue: unknown
+  comparisonValue: unknown
+  currentUser: Omit<CurrentUser, 'role'> | null
+  hidden: boolean | StateTree<boolean> | undefined
+  readOnly: boolean | StateTree<boolean> | undefined
+  openPath: Path
+  focusPath: Path
+  presence: FormNodePresence[]
+  validation: ValidationMarker[]
+  fieldGroupState: StateTree<string> | undefined
+  collapsedPaths: StateTree<boolean> | undefined
+  collapsedFieldSets: StateTree<boolean> | undefined
+  changesOpen?: boolean
+}
 
-  if (isObjectSchemaType(field.type)) {
-    const fieldValue = parentValue?.[field.name]
-    const fieldComparisonValue = isRecord(parentComparisonValue)
-      ? parentComparisonValue?.[field.name]
-      : undefined
+export interface PrepareFormState {
```

Comment:
see later comment about this being a function with properties - wondering if it would be easier to reason about this if it was a regular object with `prepareRoot`, or `prepareDocument` or smth